### PR TITLE
Release extension

### PR DESCRIPTION
**Pull request by juliandescottes**
Overview of the changes: 
- introducing /dist subfolder to put all the built extensions
- introducing /extension subfolder to put the extension sources
- ported release documentation from ADB Helper PR at https://github.com/mozilla/adbhelper/pull/44/files
- introduced template for creating platform specific manifest.json
- introduced template for update.json
- copied sign.sh from adbhelper repository
- added release target to Makefile, extension is deployed at https://ftp.mozilla.org/pub/labs/fxos-simulator/adb-extension/

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,12 +1,13 @@
 ARCHS=linux linux64 mac64 win32
-MANIFEST=manifest.json
+MANIFEST=extension/manifest.json
 
 VERSION=`grep -Eo '"version"\:\s*"[^"]+' $(MANIFEST) | grep -Eo '[0-9].*'`
 
 define build-xpis
+	cd extension; \
```
Comment (hiikezoe) [inline Makefile]: I think we should restore the current directory after the for loop below?

pushd extension
for ...
  ...
done
popd

should work.

File: Makefile [hunk 2]
```diff
@@ -1,13 +1,20 @@
 ARCHS=linux linux64 mac64 win32
-MANIFEST=extension/manifest.json
 
-VERSION=`grep -Eo '"version"\:\s*"[^"]+' $(MANIFEST) | grep -Eo '[0-9].*'`
+ROOT_PATH=/pub/labs/fxos-simulator/adb-helper/
```
Comment (hiikezoe) [inline Makefile]: I've been wondering we need a new place for the new extension.  Now we have a new good name, devtools-adb-extension, we should use this.
 /pub/devtools/devtools-adb-extension/
?
I am not sure we can create the devtools directory on the /pub.

File: Makefile [hunk 3]
```diff
@@ -1,19 +1,22 @@
 ARCHS=linux linux64 mac64 win32
 
-ROOT_PATH=/pub/labs/fxos-simulator/adb-helper/
-ROOT_UPDATE_URL=https://ftp.mozilla.org$(ROOT_PATH)
+ADDON_NAME=adb-extension
```
Comment (hiikezoe) [inline Makefile]: EXTENSION_NAME instead, not a big deal though?

File: Makefile [hunk 4]
```diff
@@ -1,19 +1,22 @@
...
+VERSION=0.0.1
+XPI_NAME=$(ADDON_NAME)-$(VERSION)
 
-VERSION=1.0
+ROOT_PATH=pub/labs/fxos-simulator/$(ADDON_NAME)
```
Comment (hiikezoe) [inline Makefile]: Oh, adb-extension looks nice.  so the ROOT_PATH should be;

 pub/devtools/$(EXTENSION_NAME)

?

File: template-update.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{
+  "addons": {
+    "adb@mozilla.org": {
+      "updates": [
+        {
+          "version": "@@VERSION@@",
+          "update_link": "@@UPDATE_LINK@@",
+          "applications": {
+            "gecko": {
+              "strict_min_version": "63.0a1"
```
Comment (hiikezoe) [inline template-update.json]: Now nightly is already 64, isn't it?

File: Makefile [hunk 6]
```diff
@@ -24,10 +27,45 @@ define clean
 	rm -f **/*.xpi
 endef
 
-# default target, clean previous XPIs and build new ones.
+define release
+	cd dist;                                                                     \
```
Comment (hiikezoe) [inline Makefile]: Again, should we restore the current directory?

File: Makefile [hunk 7]
```diff
@@ -24,10 +27,45 @@ define clean
...
+		aws s3 cp                                                                  \
+			$(XPI_NAME)-$$arch.xpi                                                   \
+			$(S3_BASE_URL)/$$arch/$(XPI_NAME)-$$arch.xpi;                            \
+                                                                               \
+		echo "[release-$$arch] Copy to 'latest' .xpi";                             \
+		aws s3 cp                                                                  \
+			$(S3_BASE_URL)/$$arch/$(XPI_NAME)-$$arch.xpi                             \
+			$(S3_BASE_URL)/$$arch/adb-extension-latest-$$arch.xpi;                   \
+                                                                               \
+		echo "[release-$$arch] Create update.json";                                \
```
Comment (hiikezoe) [inline Makefile]: Creating update.json and uploading it should be outside this loop.

File: RELEASE.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
+# Release a new version
+
+1) Update add-on version
```
Comment (hiikezoe) [inline RELEASE.md]: Though I think we should call this add-on 'extension',  is it intentional to use 'add-on's in this document?
If you are convinced that we should use 'add-on', then I am fine with that. :)

**Comment by hiikezoe** [comment]
Looks pretty good!  I did comment some minor things.  Please feel free to merge these changes with the minor changes.
Thank you!!

File: Makefile (hunk 7 above)
Comment (juliandescottes) [reply Makefile]: There is one update.json file per platform, so I think it should remain in the loop?

File: Makefile (hunk 4 above)
Comment (juliandescottes) [reply Makefile]: I don't have access to pub/* , but I actually have access to pub/labs/*, so I should be able to make that pub/labs/devtools at least.

File: Makefile (hunk 7 above)
Comment (hiikezoe) [reply Makefile]: Oh I see. I missed the part that the json is uploaded in different directories.

File: Makefile (hunk 1 above)
Comment (juliandescottes) [reply Makefile]: addressed in last commit

File: Makefile (hunk 4 above)
Comment (juliandescottes) [reply Makefile]: Done for `pub/labs/devtools`, will need to update the WebIDE patch accordingly

File: Makefile (hunk 3 above)
Comment (juliandescottes) [reply Makefile]: done

File: Makefile (hunk 6 above)
Comment (juliandescottes) [reply Makefile]: done in last commit

File: template-update.json (hunk 5 above)
Comment (juliandescottes) [reply template-update.json]: done

File: RELEASE.md (hunk 8 above)
Comment (juliandescottes) [reply RELEASE.md]: Mostly because I copied it from another place :)

**Comment by juliandescottes** [comment]
Changed the formatting of the Makefile, the "\\" padded with tabs were displayed terribly in Github's diff view, so we can't really use this.

File: extension/template-manifest.json [hunk 9]
```diff
@@ -2,13 +2,13 @@
...
   "name": "ADB binary provider",
   "author": "Mozilla & Android Open Source Project",
-  "version": "1.0",
-  "description": "An extension providing adb blobs for connecting to Firefox for Android",
+  "version": "@@VERSION@@",
+  "description": "An extension providing ADB blobs for connecting to Firefox for Android",
   "applications": {
     "gecko": {
       "id": "adb@mozilla.org",
       "strict_min_version": "63.0a1",
```
Comment (ochameau) [inline extension/template-manifest.json]: Should this be updated to match the min version set in update.json? (i.e. 64.0a1)

File: Makefile [hunk 10]
```diff
@@ -1,18 +1,69 @@
...
+		echo "[release-$$arch] Delete temporary manifest.json"; \
+		rm manifest.json; \
+	done; \
+	popd
+endef
+
+define clean
+	echo "Remove previous xpi files"; \
+	rm -f *.xpi \
+	rm -f **/*.xpi
```
Comment (ochameau) [inline Makefile]: Should this only be: `rm dist/*.xpi` ?

File: Makefile [hunk 11]
```diff
@@ -1,18 +1,69 @@
 ARCHS=linux linux64 mac64 win32
-MANIFEST=manifest.json
 
-VERSION=`grep -Eo '"version"\:\s*"[^"]+' $(MANIFEST) | grep -Eo '[0-9].*'`
+EXTENSION_NAME=adb-extension
+VERSION=0.0.2
```
Comment (ochameau) [inline Makefile]: Should the version have "pre" suffix here?

**Review by ochameau — APPROVED** [review]
Looks good to me!
The ultimate review is on your side, using this and verify that uploaded files are correct :)
If you want, we can also review them once you get to use this script in its final version.

File: extension/template-manifest.json (hunk 9 above)
Comment (juliandescottes) [reply extension/template-manifest.json]: Good catch, will fix this!

File: Makefile (hunk 10 above)
Comment (juliandescottes) [reply Makefile]: Yes that should be enough now, I still had old xpis in my repo (generated under the root) but no need to keep this here/

File: Makefile (hunk 11 above)
Comment (juliandescottes) [reply Makefile]: Yes good point I should modify that here and then do a first release PR.

**Comment by juliandescottes** [comment]
> The ultimate review is on your side, using this and verify that uploaded files are correct :)
If you want, we can also review them once you get to use this script in its final version.

Actually, in Nightly WebIDE we already use the 0.0.2 version I released quietly while testing this ;)

**Pull request merged**